### PR TITLE
[Auto Parallel] Add spmd rule No.4、13 for (batch_norm,sync_batch_norm) and their backward ops. 

### PR DESCRIPTION
**Pull request by Glencsa**
### PR Category

Auto Parallel

### PR Types

New features

### Description

- 【开源任务】算子切分推导规则开发，支持更多模型使用自动并行，简化更多用户的分布式开发成本。
- [No.4 batch_norm](https://github.com/PaddlePaddle/Paddle/issues/72415#issuecomment-2899735426)
[No.13 sync_batch_norm](https://github.com/PaddlePaddle/Paddle/issues/72415#issuecomment-2911628674)
- 将除了做batch_norm以外的维度全部强制为Replicated

### CODE REVIEW

File: test/cpp/auto_parallel/spmd_rule_test.cc [hunk 1]
```diff
@@ -2614,7 +2785,7 @@ TEST(Topk, Ctor) {
 
   // test forward
   // axis = 1
-  // [0, 1, -1] -> [0, -1, -1], [0, -1, -1]
+  // [0, -1, -1, 1],[-1],[-1],[-1],[-1] ->[-1 , -1, -1, 1],[1],[1],[1],[1],[1]
```
Comment (jeff41404) [inline test/cpp/auto_parallel/spmd_rule_test.cc]: should not modify this annotation?

File: paddle/phi/infermeta/spmd_rules/batch_norm.cc [hunk 2]
```diff
@@ -0,0 +1,731 @@
...
+                            const DistMetaTensor& variance,
+                            const DistMetaTensor& scale,
+                            const DistMetaTensor& bias,
+                            const bool is_test,
+                            const float momentum,
+                            const float epsilon,
+                            const std::string data_format,
+                            const bool use_global_stats,
+                            const bool trainable_statistics) {
+  return BatchNormInferSpmdBase(x, mean, variance, scale, bias);
```
Comment (jeff41404) [inline paddle/phi/infermeta/spmd_rules/batch_norm.cc]: do we need parameter of `data_format` in `BatchNormInferSpmdBase`? 
if user pass `data_format="NHWC" or "NLC"`, will right?

File: paddle/phi/infermeta/spmd_rules/batch_norm.cc [hunk 3]
```diff
@@ -0,0 +1,731 @@
...
+                                const DistMetaTensor& saved_variance,
+                                const DistMetaTensor& reserve_space,
+                                const DistMetaTensor& out_grad,
+                                const float momentum,
+                                const float epsilon,
+                                const std::string data_format,
+                                const bool is_test,
+                                const bool use_global_stats,
+                                const bool trainable_statistics) {
+  return BatchNormGradInferSpmdBase(x,
```
Comment (jeff41404) [inline paddle/phi/infermeta/spmd_rules/batch_norm.cc]: same issue of `data_format ` as in `BatchNormInferSpmdBase`

File: paddle/phi/ops/yaml/ops.yaml [hunk 4]
```diff
@@ -5056,6 +5056,7 @@
   output : Tensor(out), Tensor(mean_out), Tensor(variance_out), Tensor(saved_mean), Tensor(saved_variance), Tensor(reserve_space)
   infer_meta :
     func : BatchNormInferMeta
+    spmd_rule : SyncBatchNormInferSpmd
```
Comment (jeff41404) [inline paddle/phi/ops/yaml/ops.yaml]: the operator of `sync_batch_norm_` is used for manual parallelism, and the implementation of operator includes communication, not just a calculation operator. should have spmd rule?

File: paddle/phi/infermeta/spmd_rules/batch_norm.cc [hunk 5]
```diff
@@ -0,0 +1,731 @@
...
+          << str_join(mean_dist_attr.dims_mapping()) << "]";
+  VLOG(4) << "Variance_out dims mapping: ["
+          << str_join(variance_dist_attr.dims_mapping()) << "]";
+  VLOG(4) << "Saved_mean dims mapping: ["
+          << str_join(mean_dist_attr.dims_mapping()) << "]";
+  VLOG(4) << "Saved_variance dims mapping: ["
+          << str_join(variance_dist_attr.dims_mapping()) << "]";
+  VLOG(4) << "Reserve_space dims mapping: ["
+          << str_join(reserve_space_dist_attr.dims_mapping()) << "]";
+  VLOG(4) << std::endl;
```
Comment (jeff41404) [inline paddle/phi/infermeta/spmd_rules/batch_norm.cc]: shall we use macro `LOG_SPMD_INPUT` or `LOG_SPMD_OUTPUT` to simplify log code

File: paddle/phi/infermeta/spmd_rules/batch_norm.cc [hunk 6]
```diff
@@ -0,0 +1,731 @@
...
+          << "src_dims_mapping: [" << str_join(scale.dist_attr().dims_mapping())
+          << "] "
+          << "dst_dims_mapping: [" << str_join(scale_attr_dst.dims_mapping())
+          << "]";
+  VLOG(4) << "Bias"
+          << " shape: [" << str_join(bias_shape) << "] "
+          << "src_dims_mapping: [" << str_join(bias.dist_attr().dims_mapping())
+          << "] "
+          << "dst_dims_mapping: [" << str_join(bias_attr_dst.dims_mapping())
+          << "]";
```
Comment (jeff41404) [inline paddle/phi/infermeta/spmd_rules/batch_norm.cc]: shall we use macro `LOG_SPMD_INPUT` or `LOG_SPMD_OUTPUT` to simplify log code

File: test/cpp/auto_parallel/spmd_rule_test.cc (hunk 1 above)
Comment (Glencsa) [reply test/cpp/auto_parallel/spmd_rule_test.cc]: Thanks for your notice, I will revert my changes in next commit.

File: paddle/phi/ops/yaml/ops.yaml (hunk 4 above)
Comment (Glencsa) [reply paddle/phi/ops/yaml/ops.yaml]: Thanks, I think you are right, the  operator of **sync_batch_norm_** cause different GPUs have different batch, and their **mean** and **variance** on device should be communication, the tensor can not be sharded. I will remove the spmd_rule for **sync_batch_norm_** in next commit.

File: paddle/phi/infermeta/spmd_rules/batch_norm.cc (hunk 5 above)
Comment (Glencsa) [reply paddle/phi/infermeta/spmd_rules/batch_norm.cc]: Thanks, I will use it to simplify log code in next commit.

File: paddle/phi/infermeta/spmd_rules/batch_norm.cc (hunk 6 above)
Comment (Glencsa) [reply paddle/phi/infermeta/spmd_rules/batch_norm.cc]: Thanks, I will use it to simplify log code in next commit.

File: paddle/phi/infermeta/spmd_rules/batch_norm.cc (hunk 2 above)
Comment (Glencsa) [reply paddle/phi/infermeta/spmd_rules/batch_norm.cc]: Thanks, I have fit all the **data_format** status in the new commit.

File: paddle/phi/infermeta/spmd_rules/batch_norm.cc (hunk 3 above)
Comment (Glencsa) [reply paddle/phi/infermeta/spmd_rules/batch_norm.cc]: Thanks, I have fit all the data_format status in the new commit.

**Review by jeff41404 — APPROVED** [review]
LGTM
